docs/04-reference/winglang-spec.md

```diff
@@ -172,45 +172,287 @@ inflight (arg1: <type1>, arg2: <type2>, ...) => <type>
 
-A special data type is available in Wing called `Struct`. Not to be confused
-with definable structs which are outlined in their own section below. This type
-represents an arbitrary JSON value. This type can currently by assigned to and
-from `any`. During the assignment, the compiler internally knows about the
-length of this type and attempts to read it as either a JSON string or buffer.
+Wing has a data type called `Json`. This type represents an immutable arbitrary [JSON
+value](https://www.json.org/json-en.html) (primitives, arrays and objects).
 
-Since casting is not allowed in Wing, `Struct` data type is `any`'s interface
-with the outside world. `Struct` can be casted back to `any` at any time.  
-All structs can be casted to `Struct` and back to `any` at any time as well.  
-`Struct` cannot be casted to any other type.
+##### 1.1.4.1 Definition
 
-"Struct" is immutable by design. Meaning that once parsed from its `any` buffer,
-it cannot be modified anymore.
+Literals can be defined using the `Json` type initializers:
 
-> ```TS
-> // assuming fetch_some_data_with_jsii() returns "any":
-> let data: Struct = await fetch_some_data_with_jsii();
-> print(data[0]["prop-1"].id);
-> print(data.name);
-> ```
-  
-<details><summary>Equivalent TypeScript Code</summary>
+```js
+let json_string  = Json "hello";
+let json_number  = Json 123;
+let json_bool    = Json true;
+let json_array   = Json [ 1, 2, 3 ];
+let json_obj     = Json { boom: 123 };
+let json_mut_obj = MutJson {
+  hello: 123, 
+  world: [ 1, "cat", 3 ], // <-- heterogenous array
+  boom: { hello: 1233 }
+};
+```
 
-> ```TS
-> const data: any = Object.freeze(await (async () => {
->   try {
->     // deep copy all enumerable properties of the host object
->     return JSON.parse(JSON.stringify(await fetch_some_data_with_jsii()));
->   } catch(err) {
->     throw new Error(`Failed to parse JSON from any: ${err.message}`);
->   }
-> })());
-> console.log(data[0]["prop-1"].id);
-> console.log(data.name);
-> ```
+The type of elements within objects and arrays is still `Json`.
 
-</details>
+To access a field within an object, use the `.` notation:
+
+```js
+let boom: Json = json_obj.boom;
+```
+
+Trying to access a non-existent field will fail at runtime. For example:
+
+```js
+print(json_obj.boom.dude.world);
+// RUNTIME ERROR: Uncaught TypeError: Cannot read properties of undefined (reading 'world')
+```
+
+To access an array element, use the `[]` notation:
+
+```js
+let item2: Json = json_array[2];
+```
+
+It is also possible to assign the native `str`, `num`, `bool` and `Array<T>` values and they will
+implicitly be casted to `Json`:
+
+```js
+let my_str: str = "hello";
+let my_num: num = 183;
+let my_bool: bool = true;
+let my_arr: Array<num> = [1,2,3];
+
+let json_string = Json { 
+  a: my_string,
+  b: my_num,
+  c: my_bool,
+  d: my_arr
+};
+```
+
+##### 1.1.4.2. Parsing
+
+To convert a `Json` to a strong-type, use the `parse_json()` static method on the relevant target
+type:
```

Comment:
I think there are two kinds of "parsing" that we are interested in:
1. casting Json to structured types
2. parsing raw strings into Json

I wonder if calling (1) "parsing" could be confusing? Also I think (2) might be missing from the spec

Reply:
Yes, I am really not in love with this name.

How about `from_json`?

```js
let s: str = str.from_json(my_json.foo.bar.baz);
```

This is also aligned with the factory method pattern (CDK uses `fromXxx` for all factory methods). I know @yoav-steinberg has some cool ideas on formalizing factory methods, which means that if we have 1st class support for them then those `from_json` methods will be included.

Reply:
OK, changed to `from_json` and added `Json.parse(s): Json` for the 2nd use case. Dah!